### ISSUE TITLE
Access UObject Thumbnail Without Loading It (FAssetData?)

### ISSUE THREAD

**Issue by ElviraHT** [20tab/UnrealEnginePython#254]
Heya, i noticed there is a dedicated thumbnail API and that's fantastic, but both the examples and the docs don't mention whether or not it's possible to access an Asset Thumbnail simply by using the FAssetData, so that it isn't required to load the object.

This comes in very handy when you want to load tons of assets in a thumbnail viewer but you can't load all of them in RAM!

How would i accomplish this, is it possible?
Since the thumbnails are cached and the Editor displays them without loading any object, i guess it is, but how would i do it?

**Comment by rdeioris** [comment]
It looks like FAssetThumbnail is able to get the gfx from the FAssetData to draw it on a slate widget. I need to spot its code to check if it is truly able to do it without loading the UObject

**Comment by ElviraHT** [comment]
Okay, so even though i don't know anything about C++ or how the Engine works (i barely know what an asset registry is), i did some digging around because i don't like standing here doing nothing while someone else has to do all the work, so let's see if what i found can be of some help, sorry if it isn't but at least i tried xD



Looking at the **AssetThumbnail.cpp** (where it seems the **FAssetThumbnail** is defined), there are 2 interesting functions:
1) The one mentioned in the docs, **MakeThumbnailWidget**, which given a **FAssetThumbnailConfig** (to specify the size of the widget) returns a widget ready to be rendered in the SlateUI through the **FAssetThumbnailPool** it looks like?
This widget could be used right away to render into the UI (can this be used in game? because that's where i need to render the thumbnails) or you could extract the texture out of it, using maybe some of the code used in the example thumbnail viewer.
Of course i don't know anything about how textures are rendering or how the slate ui works, so i just threw this your way because if it is possible to extract or use the result as is you most likely know how to do it.
2) There is an interesting function which is not mentioned in the docs, which i found by digging in the Source: **FAssetThumbnail::GetViewportRenderTargetTexture()**
Of course i'm not exactly sure about what it does, but it looks like it can either:
  **A)** Access the **FAssetThumbnailPool** used by the SlateUI to extract a Texture out of it (it returns a Texture, which is why it looks useful), which could be very useful when combined with the option above, the **MakeThumbnailWidget**, in case this is the fastest way to get something usable out of the Widget.
  **B)** Or it can access the FAssetData's thumbnail as a Texture right away, which would indeed be exactly what we are looking for.

Here is the snippet of code i'm talking about, found in _Source/Editor/UnrealEd/Private/AssetThumbnail.cpp_:
```
FSlateShaderResource* FAssetThumbnail::GetViewportRenderTargetTexture() const
{
	FSlateTexture2DRHIRef* Texture = NULL;
	if ( ThumbnailPool.IsValid() )
	{
		Texture = ThumbnailPool.Pin()->AccessTexture( AssetData, Width, Height );
	}
	if( !Texture || !Texture->IsValid() )
	{
		return NULL;
	}

	return Texture;
}
```


That said, i also stumbled on something that might be the easiest solution: looks like in the Source there are some **ThumbnailTools**, which, looking at the thumbnail api here, were already used in the exposed python api.
Here are some interesting functions (in particular, the topmost one looks very interesting and might well be the function we are looking for, because it returns a FObjectThumbnail by simply giving the FullName):
```
/** Searches for an object's thumbnail in memory and returns it if found */
	UNREALED_API const FObjectThumbnail* FindCachedThumbnail( const FString& InFullName );

	/** Loads thumbnails from the specified package file name */
	UNREALED_API bool LoadThumbnailsFromPackage( const FString& InPackageFileName, const TSet<FName>& InObjectFullNames, FThumbnailMap& InOutThumbnails );

	/** Loads thumbnails from a package unless they're already cached in that package's thumbnail map */
	UNREALED_API bool ConditionallyLoadThumbnailsFromPackage( const FString& InPackageFileName, const TSet< FName >& InObjectFullNames, FThumbnailMap& InOutThumbnails );
```



These are all the observations i could make on what i found.
Obviously, being limited to observation since i have no knowledge about this stuff, this is unfortunately the extent to which i was able to push them.

Nevertheless, i hope i could be of some help in lessening your workload and making it easier to implement this feature!

**Comment by rdeioris** [comment]
Well, thank you a lot, you did 3/4 of the job :) I think i should be able to come up with a solution in the next few days

**Comment by rdeioris** [comment]
Hi, added an api for FAssetData based on your infos (thanks again):

```python
import unreal_engine as ue
from unreal_engine import FARFilter

_filter = FARFilter()
_filter.class_names = ['StaticMesh']

# when passing True to the second argument of get_assets_by_filter(), you will get FAssetData instead of UObject
for asset_data in ue.get_assets_by_filter(_filter, True):
    has_custom_thumbnail = asset_data.has_custom_thumbnail()
    has_cached_thumbnail = asset_data.has_cached_thumbnail()
    try:
        thumbnail = asset_data.get_thumbnail()
    except:
        thumbnail = None
    ue.log('Asset: {0} Loaded: {1} CustomThumbnail: {2} CachedThumbnail: {3} Thumbnail: {4}'.format(asset_data.object_path, asset_data.is_asset_loaded(), has_custom_thumbnail, has_cached_thumbnail, thumbnail)) 
```

**Comment by ElviraHT** [comment]
Awesome, and no problem!

If i may suggest, exposing a get_full_name() for FAssetData could be very handy, because the example thumbnail viewer, for instance, requires it.
Of course it's easy to set up a "%s %s" % (fassetdata.asset_class, fassetdata.object_path) to get the full name, but since we are working with python the lazy get_full_name() could be handy xD
It looks like it would be as simple as `*self->asset_data.GetFullName()`, so that might be something interesting to add to the ue_PyFAssetData_getseters[].
Just throwing this suggestion your way since it's very fast to implement and should make things "more pythonic" xD


**Also, by the way, did you successfully access the thumbnail maps in your tests, when running the method to check if it has a cached thumbnail?**
Because this is wierd, one would think that the thumbnail maps become available after the package is loaded since loading them every time from a Upackage is expensive.

**But, most importantly, how can i pass an array as a single argument when calling a blueprint function/event?**
I know how to call a blueprint function, but i don't know whether or not python lists are automagically converted to arrays.
If not, how can i make an array to pass as an argument to a blueprint function/event?

**Looks like it's possible to create a struct (importing TerrificStruct?), but if i pass a python list of struct objects as arguments to an event, will it work?**

**Comment by rdeioris** [comment]
Yes, it looks like putting the thumbnail in the cache must be manually done. I will probably add another function for doing it. Lists are automatically converted to UE arrays, and yes, once you import a struct you can instantiate it as a python class (just remember they are passed by value that is a bit counter-intuitive in python)

**Comment by ElviraHT** [comment]
Oh, yeah adding that function should be fairly straightforward since it's just a matter of caching the thumbnail obtained with fassetdata.get_thumbnail().
Which means that all you have to do is call this:
```
/**
	 * Caches a thumbnail into a package's thumbnail map.
	 *
	 * @param	ObjectFullName	the full name for the object to associate with the thumbnail
	 * @param	Thumbnail		the thumbnail to cache; specify NULL to remove the current cached thumbnail
	 * @param	DestPackage		the package that will hold the cached thumbnail
	 *
	 * @return	pointer to the thumbnail data that was cached into the package
	 */
	UNREALED_API FObjectThumbnail* CacheThumbnail( const FString& ObjectFullName, FObjectThumbnail* Thumbnail, UPackage* DestPackage );

```

**However, i think this one should do the job right away, the only caveat being that it has to re render the obj thumbnail and then it caches it:**
```
/** Generates a thumbnail for the specified object and caches it. Used to generate the thumbnail that gets saved in a package. */
	UNREALED_API FObjectThumbnail* GenerateThumbnailForObjectToSaveToDisk( UObject* InObject );
```

Everything worked perfectly with the array and the struct thing, now only one thing left.

How do i handle Textures?
In particular, i saw in the docs somewhere that the transient textures cannot be returned because they are not recognized by ue, and in fact it crashes when attempting to do so.

The methods i found by looking in the pyue source and api seem to be pretty much two:
1) create a texture with .create_texture() (not even sure i can do this while in game since there is the # WITH Editor or something, which i guess means it works only in editor?)
2) save the transient texture with save_package()

But they would both mean i have to save a temp texture somewhere, use it and then delete it.
I guess there should be a better approach i'm not aware of, so that i can create a texture2d, pass it to a blueprint function in a struct and then use it without having to save it temporarily to the disk?

The thumbnail viewer example works great when not in game, but in game it doesn't work.
**Isn't there any similar approach to pass a Texture2D in a struct to a blueprint function without having to save it?**

**Comment by muzaheed57** [comment]
hi.

were u able to do it ? 

i have done almost everything as you suggested. but i am stuck in the middle. i wss not able to geab the correct image data.
